### DISCUSSION TITLE
v1.3.0-rc.0 ready for testing

### DISCUSSION THREAD

**Discussion by ukutaht** [plausible/analytics#901] [Anything Else]
Find the changelog and installation instructions here: https://github.com/plausible/analytics/releases/tag/v1.3.0-rc.0

If you do decide to try out the release candidate, please leave a message here. I will address any issues and if everything seems smooth I will create a proper release.

**Comment by Vigasaurus** [comment]
Just deployed 1.3, and am getting
```
** (Clickhousex.Error) Code: 70, e.displayText() = DB::Exception: Unexpected inf or nan to integer conversion: while executing 'FUNCTION toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100)) :: 2) -> toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100))) UInt32 : 1' (version 21.3.4.25 (official build))
```
from /main-graph (throwing 500)

I guess it's possible some part of my data is screwed up there, but everything in the site I was testing was generated directly from a sent event. Seemingly only main-graph is throwing this. (It's seemingly only happening when data gets sampled, not on my smaller sites, and mostly only on periods longer than 30d on my medium-large site)

Docker log for the same:
```
plausible_1            | 00:35:40.211 [error] #PID<0.4188.0> running PlausibleWeb.Endpoint (connection #PID<0.4187.0>, stream id 1) terminated        
plausible_1            | Server: localhost:8000 (http)                                                                                                
plausible_1            | Request: GET /api/stats/my.site/main-graph?period=30d&date=2021-04-01&filters=%7B%7D                    
plausible_1            | ** (exit) an exception was raised:                                                                                           
plausible_1            |     ** (Clickhousex.Error) Code: 70, e.displayText() = DB::Exception: Unexpected inf or nan to integer conversion: while executing 'FUNCTION toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100)) :: 2) -> toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100))) UInt32 : 1' (version 21.3.4.25 (official build))                                                                
plausible_1            |                                                                                                                              
plausible_1            |         (ecto_sql 3.5.3) lib/ecto/adapters/sql.ex:751: Ecto.Adapters.SQL.raise_sql_call_error/1                              
plausible_1            |         (ecto_sql 3.5.3) lib/ecto/adapters/sql.ex:684: Ecto.Adapters.SQL.execute/5                                           
plausible_1            |         (ecto 3.5.5) lib/ecto/repo/queryable.ex:229: Ecto.Repo.Queryable.execute/4                                           
plausible_1            |         (ecto 3.5.5) lib/ecto/repo/queryable.ex:17: Ecto.Repo.Queryable.all/3                                                
plausible_1            |         (ecto 3.5.5) lib/ecto/repo/queryable.ex:149: Ecto.Repo.Queryable.one/3                                               
plausible_1            |         (plausible 0.0.1) lib/plausible/stats/clickhouse.ex:137: Plausible.Stats.Clickhouse.bounce_rate/2                    plausible_1            |         (plausible 0.0.1) lib/plausible_web/controllers/api/stats_controller.ex:92: PlausibleWeb.Api.StatsController.fetch_top_stats/2                                                                                                                                             
plausible_1            |         (plausible 0.0.1) lib/plausible_web/controllers/api/stats_controller.ex:13: PlausibleWeb.Api.StatsController.main_graph/2
```

**Reply by ukutaht** [reply]
Thanks for testing and reporting.

I can't reproduce this in any way at all. Tried even stuff like this in the main graph query:
```
SELECT toUInt32(round(sum(NaN))) FROM ...
```

It certainly looks like something in the [bounce rate calculation](https://github.com/plausible/analytics/blob/master/lib/plausible/stats/fragments.ex#L20) is giving you NaN or Inf. If you have time to look at it, could you take the failing SQL query, run it in isolation and see what's returning NaN or Inf in the bounce rate query? Like I said I've tried to break it in every way but I can't seem to break it.

Another thought I had... what Clickhouse version are you using? Maybe you're on a newer version that errors with NaN result whereas in the old version it was converted to 0 or something like that.

**Reply by Vigasaurus** [reply]
Will do a bit of testing and get back to you shortly (in a few hours, probably), but in short, I'm just using the `latest` docker tag for clickhouse, and always have been. I'll try the queries in isolation a bit, but the solution should be pretty trivial I believe, the `toUInt32` should just be a `toUInt32OrZero` and I believe CH will gracefully zero it. But I'll test it with my dataset and will report back.

**Reply by f-bader** [reply]
Hi @ukutaht, I get a similar error message to the one of @Vigasaurus 

It only affects queries where the recorded data includes February 2021, all other month work great. I do not have data before February 2021. Maybe it's something to do with February having just 28 days?

As soon as I roll back to v1.2.0 February works again.

```
plausible_events_db_1  | 2021.04.04 19:36:34.174950 [ 158 ] {beb57342-3184-4de8-8e33-863a37514f85} <Error> executeQuery: Code: 70, e.displayText() = DB::Exception: Unexpected inf or nan to integer conversion: while executing 'FUNCTION toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100)) :: 2) -> toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100))) UInt32 : 1' (version 21.3.4.25 (official build)) (from 172.24.0.6:56115) (in query: SELECT toUInt32(round(sum(is_bounce * sign) / sum(sign) * 100)) FROM "sessions" AS s0 SAMPLE 10000000 WHERE (s0."domain" = 'mydomainname.info') AND ((s0."timestamp" >= '2021-01-03T23:00:00') AND (s0."start" < '2021-01-31T23:00:00')) FORMAT JSONCompact), Stack trace (when copying this message, always include the lines below):
```

Complete errorlog can be found here https://controlc.com/31aa7598

**Reply by Vigasaurus** [reply]
Okay @ukutaht an update:

I'm running: `ClickHouse server version 21.3.4.25 (official build).` (`latest` on docker hub). (and it seems like quite a few others are too)

and, in this version running a toUIntxx on a NaN does definitely error out - 
![image](https://user-images.githubusercontent.com/19434157/113520026-dc8e3c80-9555-11eb-9037-2a110b02caea.png)

However, my original assumption that the `OrZero` modifier would work is also wrong - 
![image](https://user-images.githubusercontent.com/19434157/113520041-f62f8400-9555-11eb-934c-5a2f014fc995.png)

So the issue occurs if any section of a period has no data - so the sampling is really the issue - it's just if any of the queries have no sessions. See: 

![image](https://user-images.githubusercontent.com/19434157/113520224-3a6f5400-9557-11eb-936d-b19de0e7fca4.png)

and if we look for sessions in the same period, we get
![image](https://user-images.githubusercontent.com/19434157/113520241-507d1480-9557-11eb-9847-974e6c840d7f.png)

and if we run the pieces of the command in isolation we get a `nan` being passed into `toUInt32(round(...))`, which is the problematic bit. 
![image](https://user-images.githubusercontent.com/19434157/113520275-802c1c80-9557-11eb-9851-74e53f2cabc6.png)


The solution I found is something like the following, checking for NaN before passing it into `toUInt32(round(...))`
![image](https://user-images.githubusercontent.com/19434157/113520348-1d875080-9558-11eb-978f-1c4e1ae1317f.png)

but this could be cleaned up a bit, I suppose.

**Comment by bhavin192** [comment]
I upgraded to this today, the SMTP issue has been solved now (#827). As Vigasaurus pointed out. I get similar error when I switch to some date range where I don't have any data, for example 6 months. Otherwise the main graph for the site works fine for small date ranges. I haven't changed the Clickhouse version. It is the same which I have pulled when I started using Plausible. 

```
Request: GET /api/stats/geeksocket.in/main-graph?period=6mo&date=2021-04-03&filters=%7B%7D
** (exit) an exception was raised:
    ** (Clickhousex.Error) Code: 70, e.displayText() = DB::Exception: Unexpected inf or nan to integer conversion: while executing 'FUNCTION toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100)) :: 2) -> toUInt32(round(multiply(divide(sum(multiply(is_bounce, sign)), sum(sign)), 100))) UInt32 : 1' (version 21.2.2.8 (official build))

        (ecto_sql 3.5.3) lib/ecto/adapters/sql.ex:751: Ecto.Adapters.SQL.raise_sql_call_error/1
        (ecto_sql 3.5.3) lib/ecto/adapters/sql.ex:684: Ecto.Adapters.SQL.execute/5
        (ecto 3.5.5) lib/ecto/repo/queryable.ex:229: Ecto.Repo.Queryable.execute/4
        (ecto 3.5.5) lib/ecto/repo/queryable.ex:17: Ecto.Repo.Queryable.all/3
        (ecto 3.5.5) lib/ecto/repo/queryable.ex:149: Ecto.Repo.Queryable.one/3
        (plausible 0.0.1) lib/plausible/stats/clickhouse.ex:137: Plausible.Stats.Clickhouse.bounce_rate/2
        (plausible 0.0.1) lib/plausible_web/controllers/api/stats_controller.ex:92: PlausibleWeb.Api.StatsController.fetch_top_stats/2
        (plausible 0.0.1) lib/plausible_web/controllers/api/stats_controller.ex:13: PlausibleWeb.Api.StatsController.main_graph/2
```

Clickhouse version: 
```console
$ clickhouse local -V
ClickHouse client version 21.2.2.8 (official build).
```

**Comment by ukutaht** [comment]
Thanks @bhavin192 @Vigasaurus for the great feedback.

I found a function in Clickhouse `ifNotFinite` which seems to work well: https://github.com/plausible/analytics/commit/61676db4a7a5f7df14486a877c441aac76460412

Could you test out a new tag `v1.3.0-rc.1` which includes that fix and also a bunch of tiny bugfixes I made after after the last release candidate was tagged.

I will also merge https://github.com/plausible/analytics/pull/906 this week so we can have geolocation built-in with 0 config(!) for v1.3.0. But it's not included yet.

**Reply by f-bader** [reply]
Hi @ukutaht thanks. The use of `ifNotFinite` seems to work great. I do not receive any more error messages and can request all data as before.

**Reply by bhavin192** [reply]
I also started running it. The main graph is now loading fine.

**Comment by j4w8n** [comment]
I installed `v1.3.0-rc.0` today. Then installed `-rc.1` to resolve the default 30d graph issue - as already stated.

Excited to have the API on self-hosted!

**Reply by ukutaht** [reply]
Great to hear it's working OK for you.

**Comment by iSCJT** [comment]
Updated this morning to `rc.1` and all seems to be working great, love the embedded dashboard!  Thanks @ukutaht 👍

**Comment by f-bader** [comment]
On mobile clients the Goal Conversions view is a bit off.

![Screenshot_20210409-210356~2.png](https://user-images.githubusercontent.com/3834359/114228917-6e8fae00-9977-11eb-9b28-dac523588b35.png)

**Reply by Vigasaurus** [reply]
Yeah that's not really specific to the glob goals, it's just a UI issue for any long goal names. There are quite a few minor mobile UI issues - I plan to PR a fix for a lot of them at some point down the line.

**Reply by ukutaht** [reply]
That would be awesome @Vigasaurus

**Comment by ukutaht** [comment]
`v1.3.0` is released!  https://github.com/plausible/analytics/releases/tag/v1.3.0
